src/borg/_hashindex.c

```diff
@@ -248,99 +262,130 @@ count_empty(HashIndex *index)
 }
 
 /* Public API */
+
 static HashIndex *
-hashindex_read(const char *path)
+hashindex_read(PyObject *file)
 {
-    FILE *fd;
-    off_t length, buckets_length, bytes_read;
-    HashHeader header;
+    Py_ssize_t length, buckets_length, bytes_read;
+    HashHeader *header;
     HashIndex *index = NULL;
 
-    if((fd = fopen(path, "rb")) == NULL) {
-        EPRINTF_PATH(path, "fopen for reading failed");
-        return NULL;
-    }
-    bytes_read = fread(&header, 1, sizeof(HashHeader), fd);
-    if(bytes_read != sizeof(HashHeader)) {
-        if(ferror(fd)) {
-            EPRINTF_PATH(path, "fread header failed (expected %ju, got %ju)",
-                         (uintmax_t) sizeof(HashHeader), (uintmax_t) bytes_read);
-        }
-        else {
-            EPRINTF_MSG_PATH(path, "fread header failed (expected %ju, got %ju)",
-                             (uintmax_t) sizeof(HashHeader), (uintmax_t) bytes_read);
-        }
-        goto fail;
+    PyObject *header_bytes = PyObject_CallMethod(file, "read", "n", sizeof(HashHeader));
+    if(!header_bytes) {
+        /* Python exception occurred; just return */
+        return 0;
```

Comment:
maybe there could be a "goto fail" jumping to the "return index" (which is NULL then)"?